### DISCUSSION TITLE
Describe Logic/Rules for changing class hierarchy?

### DISCUSSION THREAD

**Discussion by lapidus** [geldata/gel#7023] [Q&A] [answered]
I’m curious what happens when you radically change the class (or “object type”) hierarchy?

Suppose you have these classes with various properties defined at each level:

Country extends Place 
Place extends Item

If I now change so that Country extends a completely different class with no relationship to Place or Item. What will happen to:
- the class itself
- objects (instances of the class)

Is there some documentation around these “schema and class hierarchy change” scenarios and possible edge cases here? Trying to understand better the repercussions of frequent class hierarchy changes …

Thanks!

**Comment by elprans** [comment, marked as answer]
Normally changing the object type base results in an `ALTER TYPE Foo {EXTENDING NewBase; DROP EXTENDING OldBase}`. Like with all `ALTER`-s objects stay in place, but properties or links might get dropped/added depending on how the layout changed due to the parent object type changes.  In this regard, change of type hierarchy is not very different from most type alterations.

However, if the object type changes are so large so that its "similarity" to the original type is very low, the schema diff algorithm might generate a `DROP TYPE Foo; CREATE TYPE Foo { ... }` sequence, which is where all existing objects would be deleted.  This is why `edgedb migration create` asks the "did you do X" questions to confirm that it understood the diff correctly and replying `n` to a `Did you delete X` will usually result in the diff to be corrected to an `alter`.

**Comment by lapidus** [comment]
Thank you for the great answer @elprans! I marked it as answered. 

ps1.
_Overall we'd like to migrate to EdgeDB at some point, but we have a lot of concurrent and frequent schema edits, which seem to require a more 'virtual schema model' versus the EdgeDB approach which rewrites table structure. But I can really see how EdgeDB would solve a whole lot of other project's needs where schema editing is a bit more sequential and "planned"._

ps2. 
_Relating to inheritance, do you have any docs outlining your approach to the 'diamond problem', or should I dig into the code?_

**Reply by elprans** [reply]
> Relating to inheritance, do you have any docs outlining your approach to the 'diamond problem', or should I dig into the code?

EdgeDB uses the [C3 linearization algorithm](https://en.wikipedia.org/wiki/C3_linearization) when resolving types with multiple parents (same as Python).
